### ISSUE TITLE
Refine recording of source locations

### ISSUE THREAD

**Issue by t-vi** [Lightning-AI/lightning-thunder#475]
From the discussion on #40 

@IvanYashchuk : 

If I use thunder.torch.softmax why is the source location recorded as

```
thunder/torch/__init__.py:4411: 	    return _softmax(a, dim=dim, dtype=dtype)
```

and with torch.softmax it points to the function being traced:

```
<ipython-input-14-72fa81ca31be>:2: 	    return torch.softmax(a, 0)
```

@t-vi 
This is because the "symbol boundary" for softmax is _softmax apparently (the symbols are opaque functions to the JIT components, so they are the topmost frame). Note that softmax is special in how we define it in thunder.torch (to cater for the undocumented parameter).

I see three options (for a follow-up PR):
- refuse to record locations in thunder.torch (deferring to the previous (next lower) stack frame if we are),
- provide a mechanism to fudge this (mark things "internal", might also be worthwile for "impl" functions if we see them),
- do nothing.

**Comment by t-vi** [comment]
This is a good first issue for people wanting to know more about the innards of the interpreter, but do tag me for help with choosing between the options above.

**Comment by k223kim** [comment]
Hi @t-vi , 

I would like to work on this issue. To get a better understanding with this issue I have some questions and would like to discuss what options would be appropriate to this issue.
- What is the expected outcome exactly?
- I am not quite sure why the current output is an issue. Would you be able to explain why such will be an issue to developers?
I would appreciate your help!

Best,
Kaeun

**Comment by t-vi** [comment]
Hi Kaeun!

I'm excited to hear that!

- I could be as easy as modifying `get_current_user_source_location` to skip more modules as internal and adding a test to test_core (e.g. using @IvanYashchuk 's softmax example). It might be a nice opportunity to take a peek at the interpreter for a broader picture what is going on (you'll see the call stack frames appearing in that function, as we run things though our own Python implementation there).
- For the why: To the user, the call they do (as shown by the second, the torch.softmax snippet) is likely more useful as information about the origin of a call in the trace than any internals of Thunder. 

Don't hesitate to reach out (here, slack, discord...) if you have any questions!

Best regards

Thomas

**Comment by k223kim** [comment]
Hi Thomas!

Thank you so much for the fast and detailed response! Let me look into `get_current_user_source_location` in more detail and submit a draft PR. Is there a slack/discord channel I can join? It'll be awesome if we can communicate through slack/discord :) 

Best,
Kaeun

**Comment by t-vi** [comment]
We use the Lightning discord ( linked from https://lightning.ai/community ) and slack ( https://pytorch-lightning.slack.com ).

**Comment by k223kim** [comment]
Hi Thomas,

I have submitted a PR that implements your suggestions regarding source locations. Would be awesome if you can take a look!

Best,
Kaeun
